test/extensions/filters/http/composite/composite_filter_integration_test.cc

```diff
@@ -1,49 +1,131 @@
+#include "envoy/extensions/common/matching/v3/extension_matcher.pb.validate.h"
+#include "envoy/extensions/filters/http/composite/v3/composite.pb.h"
 #include "envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.pb.h"
+#include "envoy/matcher/matcher.h"
+#include "envoy/type/matcher/v3/http_inputs.pb.h"
+#include "envoy/type/matcher/v3/http_inputs.pb.validate.h"
 
+#include "source/common/http/match_delegate/config.h"
+
+#include "test/common/http/common.h"
+#include "test/integration/filters/add_body_filter.pb.h"
+#include "test/integration/filters/set_response_code_filter_config.pb.h"
 #include "test/integration/http_integration.h"
 #include "test/mocks/http/mocks.h"
 #include "test/test_common/utility.h"
 
+#include "absl/strings/str_format.h"
 #include "gtest/gtest.h"
 
 namespace Envoy {
 
+namespace {
+
+using envoy::config::route::v3::Route;
+using envoy::config::route::v3::VirtualHost;
+using envoy::extensions::common::matching::v3::ExtensionWithMatcherPerRoute;
+using envoy::extensions::filters::http::composite::v3::ExecuteFilterAction;
+using Envoy::Protobuf::MapPair;
+using Envoy::ProtobufWkt::Any;
+using envoy::type::matcher::v3::HttpRequestHeaderMatchInput;
+using test::integration::filters::AddBodyFilterConfig;
+using test::integration::filters::SetResponseCodeFilterConfig;
+using xds::type::matcher::v3::Matcher_OnMatch;
+
 class CompositeFilterIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
                                        public HttpIntegrationTest {
 public:
   CompositeFilterIntegrationTest() : HttpIntegrationTest(Http::CodecType::HTTP1, GetParam()) {}
 
-  void initialize() override {
-    config_helper_.prependFilter(R"EOF(
-  name: composite
-  typed_config:
-    "@type": type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher
-    extension_config:
-      name: composite
+  Route* getMutableRoute(VirtualHost* vh, const bool& add) {
+    if (add) {
+      return vh->add_routes();
+    }
+    return vh->mutable_routes()->Mutable(0);
+  }
```

Comment:
From a quick look, i think here might be the problem. 
You will have two routes when `add` is true. And based on the config you shared, you wrap two per_filter_config in those two routes separately.

First, for our multiple filters test case, we only need to have one route. `typed_per_filter_config` in route is a map<string, [google.protobuf.Any]. So, one route can store route configs for two filters with distinct key.

Above is what we need to test first and probably is sufficient for our case.  Once that one works, we can explore multiple routes if you are interested in. One thing stands out from my quick scan is that those two routes have same prefix `/somepath`, the matching of routes is first matched one wins [link](https://github.com/envoyproxy/envoy/blob/a170b03435c4b12c0b6e46b6b9800f08dba2bdd5/api/envoy/config/route/v3/route_components.proto#L93). So this will introduce indeterministic behavior with same prefix you have.

Reply:
Ahhh ok that makes a lot of sense now. I think I made an appropriate fix then.